File: typescript/src/problems/s20_valid_parentheses.ts
/**
 * @link https://leetcode.com/problems/valid-parentheses/
 * @param s
 */
export function isValid(s: string): boolean {
  const stack: string[] = [];
  for (let i = 0; i < s.length; i++) {
    let c = s[i];
    if (c === '[' || c === '{' || c === '(') {
      stack.push(c);
      continue;
    }
    if (stack.length === 0) {
      return false;
    }
    switch (c) {
      case ']': {
        if (stack.pop() !== '[') {
          return false;
        }
        break;
      }
      case '}': {
        if (stack.pop() !== '{') {
          return false;
        }
        break;
      }
      case ')': {
        if (stack.pop() !== '(') {
          return false;
        }
        break;
      }
    }
  }
  return stack.length === 0;
}

File: rust/src/problems/s20_valid_parentheses.rs
struct Solution;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = vec![];
        for c in s.chars() {
            match c {
                '(' | '[' | '{' => stack.push(c),
                ')' => if stack.pop() != Some('(') { return false; },
                ']' => if stack.pop() != Some('[') { return false; },
                '}' => if stack.pop() != Some('{') { return false; },
                _ => ()
            }
        }
        stack.is_empty()
    }
}

struct Example {
    input: String,
    output: bool,
}

#[test]
pub fn test() {
    let examples = vec![
        Example {
            input: "()".to_string(),
            output: true,
        },
        Example {
            input: "()[]{}".to_string(),
            output: true,
        },
        Example {
            input: "(]".to_string(),
            output: false,
        },
        Example {
            input: "([)]".to_string(),
            output: false,
        },
        Example {
            input: "{{}[][[[]]]}".to_string(),
            output: true,
        },
    ];
    for example in examples {
        assert_eq!(Solution::is_valid(example.input), example.output);
    }
}
